source/UF/CumulativeHierarchy.lagda

```diff
@@ -0,0 +1,416 @@
+ordinal of set theoretic ordinals in 𝕍 (w.r.t. 𝓤) is isomorphic to the ordinal
+of ordinals in 𝓤.
+
+This file has three parts:
+(I)    Introduction of the cumulative hierarchy 𝕍 and the statement of its
+       (most general) induction principle.
+(II)   Statements and proofs of some simpler, more specialised, induction and
+       recursion principles for 𝕍.
+(III)  Basic constructions and proofs for 𝕍, i.e. the definition of set
+       membership (∈), subset relation (⊆) and proofs of ∈-extensionality and
+       ∈-induction.
+
+The cumulative hierarchy 𝕍 can be seen as a HoTT-refined version of Aczel's [2]
+type theoretic interpretation of constructive set theory and draws inspiration
+form Joyal and Moerdijk's [3] algebraic set theory.
+
+References
+----------
+
+[1] The Univalent Foundations Program
+    Homotopy Type Theory: Univalent Foundations of Mathematics
+    https://homotopytypetheory.org/book
+    Institute for Advanced Study
+    2013
+
+[2] Peter Aczel
+    The type theoretic interpretation of constructive set theory
+    In A. MacIntyre, L. Pacholski, and J. Paris (eds.) Logic Colloquium ’77
+    Volume 96 of Studies in Logic and the Foundations of Mathematics
+    Pages 55–66
+    North-Holland Publishing Company
+    1978
+    doi:10.1016/S0049-237X(08)71989-X
+
+[3] A. Joyal and I. Moerdijk
+    Algebraic set theory
+    Volume 220 of London Mathematical Society Lecture Note Series
+    Cambridge University Press
+    1995
+    doi:10.1017/CBO9780511752483
+
+\begin{code}
+
+{-# OPTIONS --without-K --exact-split --safe --auto-inline #-}
+
+open import UF.FunExt
+open import UF.Subsingletons
+open import UF.PropTrunc
+
+module UF.CumulativeHierarchy
+        (pt : propositional-truncations-exist)
+        (fe : Fun-Ext)
+        (pe : Prop-Ext)
+       where
+
+open PropositionalTruncation pt
+
+open import MLTT.Spartan
+open import UF.Base hiding (_≈_)
+open import UF.Subsingletons-FunExt
+
+_≲_ : {A : 𝓤 ̇ } {B : 𝓥 ̇ } {X : 𝓣 ̇ } → (A → X) → (B → X) → 𝓤 ⊔ 𝓥 ⊔ 𝓣 ̇
+_≲_ {𝓤} {𝓥} {𝓣} {A} {B} f g = (a : A) → ∃ b ꞉ B , g b ＝ f a
+
+-- Note that _≈_ says that f and g have equal images
+_≈_ : {A : 𝓤 ̇ } {B : 𝓥 ̇ } {X : 𝓣 ̇ } → (A → X) → (B → X) → 𝓤 ⊔ 𝓥 ⊔ 𝓣 ̇
+f ≈ g = f ≲ g × g ≲ f
+
+\end{code}
+
+Part I
+------
+
+Introduction of the cumulative hierarchy 𝕍 and the statement of its (most
+general) induction principle.
+
+See Section 10.5 of the HoTT Book [1] for more of an explanation regarding the
+induction principle of 𝕍.
+
+For comparison, the higher inductive type (HIT) presentation of 𝕍 (w.r.t. 𝓤) is:
+  ∙ For every A : 𝓤 and f : A → 𝕍, we have an element 𝕍-set f : 𝕍
+  ∙ For every A, B : 𝓤, f : A → 𝕍 and g : B → 𝕍, if f ≈ g, then 𝕍-set f ＝ 𝕍-set g
+  ∙ 𝕍 is set-truncated: for every x, y : 𝕍 and p, q : x ＝ y, we have p ＝ q.
+
+\begin{code}
+
+module _ (𝓤 : Universe) where
+
+ record cumulative-hierarchy-exists : 𝓤ω where
+  field
+   𝕍 : 𝓤 ⁺ ̇
+   𝕍-is-set : is-set 𝕍
```

Comment:
I think it is worth adding a remark here, for set theorists. Of course, 𝕍 is a proper class not a set. But also, of course, line 102 is saying that it is a (large) set in the sense of HoTT/UF. I wonder whether we should define `is-large-set`, and use that instead. Or else you should add a comment right at line 102. I probably prefer the first solution.

Reply:
Nice suggestion, thanks! (I went with the first solution.)